# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/workers/loader/WorkerModuleLoader.cpp
# Commit: 474668f6615e
# Full Hash: 474668f6615eafb7bc9a94b5bcf6aada66ea5d99
# Author: Yulia Startsev <ystartsev@mozilla.com>
# Date: 2023-01-19 04:46:07
# Regressor Bug: 1247687
# File Overlap Count: 1
# Description:
#   Bug 1247687 - Implement initial Static Module Loading for Workers; r=jonco
#   
#   This is the first pass of getting static module loading to work. This roughly implements
#   https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-worklet/module-worker-script-graph --
#   without some of the settings objects correctly set.
# ==============================================================================

diff -r 15eb9daca661 -r 474668f6615e dom/workers/loader/WorkerModuleLoader.cpp
--- a/dom/workers/loader/WorkerModuleLoader.cpp	Wed Jan 18 13:46:30 2023 +0000
+++ b/dom/workers/loader/WorkerModuleLoader.cpp	Wed Jan 18 13:46:30 2023 +0000
@@ -4,6 +4,9 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+#include "js/experimental/JSStencil.h"  // JS::Stencil, JS::CompileModuleScriptToStencil, JS::InstantiateModuleStencil
+#include "js/loader/ModuleLoadRequest.h"
+#include "mozilla/dom/WorkerLoadContext.h"
 #include "mozilla/dom/workerinternals/ScriptLoader.h"
 #include "WorkerModuleLoader.h"
 
@@ -31,7 +34,18 @@
 
 already_AddRefed<ModuleLoadRequest> WorkerModuleLoader::CreateStaticImport(
     nsIURI* aURI, ModuleLoadRequest* aParent) {
-  MOZ_CRASH("Not implemented yet");
+  Maybe<ClientInfo> clientInfo = GetGlobalObject()->GetClientInfo();
+
+  RefPtr<WorkerLoadContext> loadContext =
+      new WorkerLoadContext(WorkerLoadContext::Kind::StaticImport, clientInfo);
+  RefPtr<ModuleLoadRequest> request = new ModuleLoadRequest(
+      aURI, aParent->mFetchOptions, SRIMetadata(), aParent->mURI, loadContext,
+      false, /* is top level */
+      false, /* is dynamic import */
+      this, aParent->mVisitedSet, aParent->GetRootModule());
+
+  request->mURL = request->mURI->GetSpecOrDefault();
+  return request.forget();
 }
 
 already_AddRefed<ModuleLoadRequest> WorkerModuleLoader::CreateDynamicImport(
@@ -48,19 +62,53 @@
 }
 
 nsresult WorkerModuleLoader::StartFetch(ModuleLoadRequest* aRequest) {
-  return NS_ERROR_FAILURE;
+  if (!GetScriptLoader()->DispatchLoadScript(aRequest)) {
+    return NS_ERROR_FAILURE;
+  }
+  return NS_OK;
 }
 
 nsresult WorkerModuleLoader::CompileFetchedModule(
     JSContext* aCx, JS::Handle<JSObject*> aGlobal, JS::CompileOptions& aOptions,
     ModuleLoadRequest* aRequest, JS::MutableHandle<JSObject*> aModuleScript) {
-  return NS_ERROR_FAILURE;
+  RefPtr<JS::Stencil> stencil;
+  MOZ_ASSERT(aRequest->IsTextSource());
+  MaybeSourceText maybeSource;
+  nsresult rv = aRequest->GetScriptSource(aCx, &maybeSource);
+  NS_ENSURE_SUCCESS(rv, rv);
+
+  auto compile = [&](auto& source) {
+    return JS::CompileModuleScriptToStencil(aCx, aOptions, source);
+  };
+  stencil = maybeSource.mapNonEmpty(compile);
+
+  if (!stencil) {
+    return NS_ERROR_FAILURE;
+  }
+
+  JS::InstantiateOptions instantiateOptions(aOptions);
+  aModuleScript.set(
+      JS::InstantiateModuleStencil(aCx, instantiateOptions, stencil));
+  if (!aModuleScript) {
+    return NS_ERROR_FAILURE;
+  }
+
+  return NS_OK;
 }
 
 WorkerScriptLoader* WorkerModuleLoader::GetScriptLoader() {
   return static_cast<WorkerScriptLoader*>(mLoader.get());
 }
 
-void WorkerModuleLoader::OnModuleLoadComplete(ModuleLoadRequest* aRequest) {}
+void WorkerModuleLoader::OnModuleLoadComplete(ModuleLoadRequest* aRequest) {
+  if (aRequest->IsTopLevel()) {
+    AutoJSAPI jsapi;
+    if (NS_WARN_IF(!jsapi.Init(GetGlobalObject()))) {
+      return;
+    }
+    GetScriptLoader()->MaybeMoveToLoadedList(aRequest);
+    GetScriptLoader()->ProcessPendingRequests(jsapi.cx());
+  }
+}
 
 }  // namespace mozilla::dom::workerinternals::loader